interfaces/apparmor/backend.go

```diff
@@ -63,6 +70,107 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+
+	// Create the local policy directory if it is not there.
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("cannot create snap-confine policy directory, %s", err)
+	}
+
+	// Check if NFS is mounted at or under $HOME. Because NFS is not
+	// transparent to apparmor we must alter our profile to counter that and
+	// allow snap-confine to work.
+	if nfs, err := isHomeUsingNFS(); err != nil {
+		return err
+	} else if nfs {
+		policy["generated-nfs"] = &osutil.FileState{
+			Content: []byte(nfsSnippet),
+			Mode:    0644,
+		}
+		logger.Noticef("snapd enabled NFS support, additional implicit network permissions granted")
+	}
+
+	// Ensure that generated policy is what we computed above.
+	created, removed, err := osutil.EnsureDirState(dir, glob, policy)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize snap-confine policy, %s", err)
+	}
+	if len(created) == 0 && len(removed) == 0 {
+		// If the generated policy didn't change, we're all done.
+		return nil
+	}
+
+	// If snapd is executing from the core snap the it means it has
+	// re-executed. In that case we are no longer using the copy of
+	// snap-confined from the host distribution but our own copy. We don't have
+	// to re-compile and load the updated profile as that is performed by
+	// setupSnapConfineReexec below.
+	exe, err := os.Readlink(procSelfExe)
+	if err != nil {
+		return fmt.Errorf("cannot read /proc/self/exe, %s", err)
+	}
+	if strings.HasPrefix(exe, dirs.SnapMountDir) {
+		return nil
+	}
+
+	// Reload the apparmor profile of snap-confine. This points to the main
+	// file in /etc/apparmor.d/ as that file contains include statements that
+	// load any of the files placed in /var/lib/snapd/apparmor/snap-confine.d/.
+	// We are not using apparmor.LoadProfile() because it uses other cache.
+	if output, err := exec.Command("apparmor_parser", "--replace",
+		// TODO: the name varies across distros, fix that :/
+		"--write-cache", filepath.Join(dirs.SystemApparmorDir, "usr.lib.snapd.snap-confine.real"),
+		"--cache-loc", dirs.SystemApparmorCacheDir).CombinedOutput(); err != nil {
+		return fmt.Errorf("cannot reload snap-confine apparmor profile: %v", osutil.OutputErr(output, err))
+	}
+	return nil
+}
+
+var (
+	procSelfMountInfo = mount.ProcSelfMountInfo
+)
+
+// isHomeUsingNFS returns true if /home contains or might contain NFS mounts.
+//
+// Internally /proc/self/mountinfo and /etc/fstab are interrogated (for current
+// and possible mounted filesystems).  If either of those describes NFS
+// filesystem mounted under or beneath /home/ then the return value is true.
+func isHomeUsingNFS() (bool, error) {
+	mountinfo, err := mount.LoadMountInfo(procSelfMountInfo)
+	if err != nil {
+		return false, fmt.Errorf("cannot parse /proc/self/mountinfo, %s", err)
+	}
+	for _, entry := range mountinfo {
+		if (entry.FsType == "nfs4" || entry.FsType == "nfs") && (strings.HasPrefix(entry.MountDir, "/home/") || entry.MountDir == "/home") {
```

Comment:
I know we've a deeper and historical problem with hardcoded /home, but I'm not super happy about adding another hardcoded case here. Actually, if we ever fix the problem with hardcoded /home in dirs.go, this fix will break completely and this makes me a sad panda...
Would it be possible to make it more generic and future-proof (e.g. check if any of the filesystems that the snap may potentially be interested in is NFS)? Don't we get the same problem if - say - /usr or /var is NFS-mounted?

Reply:
This was added by a request from @jdstrand - the idea is that we want to slowly open this up to more use cases. For now we just want to support /home on NFS. We plan to expand this to CIFS as well but not in arbitrary places yet.

Note that `/home` is already explicitly described in apparmor profiles and there's no easy way around it today.

Reply:
Fair enough, thanks.